x-pack/legacy/plugins/ml/public/application/routing/routes/explorer.tsx

```diff
@@ -44,111 +49,131 @@ export const explorerRoute: MlRoute = {
-
-    // Now that appState in explorerState$ is the single source of truth,
-    // subscribe to it and update the actual URL appState on changes.
-    subscriptions.add(
-      explorerService.appState$.subscribe((appStateIn: ExplorerAppState) => {
-        // appState.fetch();
-        appState.mlExplorerFilter = appStateIn.mlExplorerFilter;
-        appState.mlExplorerSwimlane = appStateIn.mlExplorerSwimlane;
-        // appState.save();
-      })
-    );
-  });
+const ExplorerUrlStateManager: FC<ExplorerUrlStateManagerProps> = ({ jobsWithTimeRange }) => {
+  const [, setAppState] = useUrlState('_a');
+  const [globalState] = useUrlState('_g');
+  const [lastRefresh, setLastRefresh] = useState(0);
 
-  subscriptions.add(subscribeAppStateToObservable(AppState, 'mlShowCharts', showCharts$, () => {}));
-  subscriptions.add(
-    subscribeAppStateToObservable(AppState, 'mlSelectInterval', interval$, () => {})
-  );
-  subscriptions.add(
-    subscribeAppStateToObservable(AppState, 'mlSelectSeverity', severity$, () => {})
-  );
+  const { jobIds } = useJobSelection(jobsWithTimeRange, getDateFormatTz());
 
-  if (globalState.time) {
-    timefilter.setTime({
-      from: globalState.time.from,
-      to: globalState.time.to,
-    });
-  }
+  const refresh = useRefresh();
+  useEffect(() => {
+    if (refresh !== undefined) {
+      setLastRefresh(refresh?.lastRefresh);
+      const activeBounds = timefilter.getActiveBounds();
+      if (activeBounds !== undefined) {
+        explorerService.setBounds(activeBounds);
+      }
+    }
+  }, [refresh?.lastRefresh]);
 
   useEffect(() => {
-    return () => {
-      subscriptions.unsubscribe();
-      unsubscribeFromGlobalState();
-    };
-  });
+    timefilter.enableTimeRangeSelector();
+    timefilter.enableAutoRefreshSelector();
+  }, []);
+
+  useEffect(() => {
+    if (globalState?.time !== undefined) {
+      timefilter.setTime({
+        from: globalState.time.from,
+        to: globalState.time.to,
+      });
+      explorerService.setBounds({
+        min: moment(globalState.time.from),
+        max: moment(globalState.time.to),
+      });
+    }
+  }, [JSON.stringify(globalState?.time)]);
+
+  useEffect(() => {
+    if (jobIds.length > 0) {
+      explorerService.updateJobSelection(jobIds);
+    } else {
+      explorerService.clearJobs();
+    }
+  }, [JSON.stringify(jobIds)]);
+
+  const [explorerData, loadExplorerData] = useExplorerData();
+  useEffect(() => {
+    if (explorerData !== undefined && Object.keys(explorerData).length > 0) {
+      explorerService.setExplorerData(explorerData);
+    }
+  }, [JSON.stringify(explorerData)]);
+
+  const explorerAppState = useObservable(explorerService.appState$);
+  useEffect(() => {
+    if (
+      explorerAppState !== undefined &&
+      explorerAppState.mlExplorerSwimlane.viewByFieldName !== undefined
+    ) {
+      setAppState(explorerAppState);
+    }
+  }, [JSON.stringify(explorerAppState)]);
```

Comment:
We want to update `appState` on any `explorerAppState` change, the check against `viewByFieldName` is just a safety measure because we don't want to store `undefined` on initialization in the URL. Since `explorerService.appState$` already does a `distinctUntilChanged(isEqual)` I'll remove the `JSON.stringify` though, good point.